dfc/proxy.go

```diff
@@ -325,50 +331,38 @@ func (p *proxyrunner) findRole() {
 		return
 	}
 
-	deadline := time.Now().Add(ctx.config.Timeout.Startup)
+	smap, lbmap := p.discoverServers(hint, time.Now().Add(ctx.config.Timeout.Startup), startupPollSleepTime)
+	lbmapLock.Lock()
+	if lbmap != nil && p.lbmap.version() < lbmap.version() {
+		p.lbmap = lbmap
+	}
+	lbmapLock.Unlock()
 
-	for {
-		u := p.smap.Join(hint)
-		smap, lbmap, done := p.discoverMaps(u)
-		if done {
-			lbmapLock.Lock()
-			if lbmap != nil && p.lbmap.version() < lbmap.version() {
-				p.lbmap = lbmap
-			}
-			lbmapLock.Unlock()
-
-			if smap != nil {
-				smapLock.Lock()
-				if smap.ProxySI.DaemonID == p.si.DaemonID {
-					glog.Infoln("This proxy is primary in found cluster map; remaining primary.")
-					p.smap = smap
-					smapLock.Unlock()
-					return
-				}
-				smapLock.Unlock()
-
-				glog.Infof("This proxy (%s) appears to be not primary in the max-versioned cluster map", p.si.DaemonID)
-				glog.Infof("Registering with the real primary %s", smap.ProxySI.DaemonID)
-				err := p.registerWithRetry(smap.ProxySI.DirectURL, ctx.config.Timeout.Default)
-				if err != nil {
-					glog.Errorf("Error registering with primary proxy %v: %v. Retrying.", smap.ProxySI.DaemonID, err)
-				} else {
-					smapLock.Lock()
-					defer smapLock.Unlock()
-					p.becomeNonPrimaryProxy()
-					p.smap = smap
-					p.setPrimaryProxy(smap.ProxySI.DaemonID, "" /* primaryToRemove */, false /* prepare */)
-					return
-				}
-			}
-		}
+	if smap == nil {
+		return
+	}
 
-		if time.Now().After(deadline) {
-			glog.Warning("findRole took too long")
-			return
-		}
+	smapLock.Lock()
+	if smap.ProxySI.DaemonID == p.si.DaemonID {
+		glog.Infoln("This proxy is primary in found cluster map; remaining primary.")
```

Comment:
I did not check: is it possible that discover returns smap with this proxy as primary but at this moment p.primary is false?

Reply:
not possible, this path is for a proxy to start as primary and trying to figuring out what is the latest map in order to decide whether it should give up the rights to be primary and down grade itself to non primary